File: othello.py
from __future__ import print_function
from functools import reduce
from copy import deepcopy
import glob

import numpy as np
import chainer
import chainer.functions as F
import chainer.links as L
from chainer import training, Variable, cuda
from chainer.training import extensions
from neural_net import MLP, Classifier


EMPTY = 0
PLAYER = 1
OPPONENT = 2
VALID = 3

DISC_MARK = {
  EMPTY: '-',
  PLAYER: 'o',
  OPPONENT: 'x',
  VALID: '.'
}

class Board():
  def __init__(self, board=None):
    if board is not None:
      self.cells = board.copy()
      return
    self.cells = [[EMPTY] * (8) for _ in range(8)]
    self[4, 5] = PLAYER
    self[5, 4] = PLAYER
    self[4, 4] = OPPONENT
    self[5, 5] = OPPONENT

  def copy(self):
    return deepcopy(self.cells)

  def __getitem__(self, disc_pos):
    if (1 <= disc_pos[0] <= 8) and (1 <= disc_pos[1] <= 8):
      return self.cells[disc_pos[1] - 1][disc_pos[0] - 1]
    return EMPTY

  def __setitem__(self, disc_pos, value):
    self.cells[disc_pos[1] - 1][disc_pos[0] - 1] = value
    return self.cells

  def __str__(self):
    rangement = [str(i + 1) + ' ' for i in range(8)]
    string = '  {}\n'.format(reduce(lambda a, e: a + e, rangement, ''))
    for i, row in enumerate(self.cells):
      string += '{} {}\n'.format(i + 1,
                  reduce(lambda a, e: a + DISC_MARK[e] + ' ', row, ''))
    return string


class Othello():
  def __init__(self):
    self.board = Board()
    self.valid_moves = []
    self.update_valid_moves()

  def make_move(self, move):
    for m in self.valid_moves:
      self.board[m] = EMPTY
    self._flip_disks(self.board, move)
    self.switch_turn()
    self.update_valid_moves()
    if not self.valid_moves:
      self.switch_turn()

  def update_valid_moves(self):
    del self.valid_moves[:]
    for row in range(1, 9):
      for col in range(1, 9):
        if self.board[row, col] == EMPTY:
          board = Board(self.board)
          if self._flip_disks(board, (row, col)):
            self.valid_moves.append((row, col))
    self.valid_moves = self._uniq(self.valid_moves)
    for m in self.valid_moves:
      self.board[m] = VALID

  def switch_turn(self):
    for row in range(1, 9):
      for col in range(1, 9):
        if self.board[row, col] == EMPTY: continue
        self.board[row, col] = OPPONENT if self.board[row, col] == PLAYER else PLAYER

  def _flip_disks(self, board, move):
    board[move] = PLAYER
    directions = {
      'right' : (9 - move[0], ( 1,  0)),
      'bottom': (9 - move[1], ( 0,  1)),
      'left'  : (move[0],     (-1,  0)),
      'top'   : (move[1],     ( 0, -1)),
      'top_left'    : (min(move[1],     move[0]),     (-1, -1)),
      'bottom_right': (min(9 - move[1], 9 - move[0]), ( 1,  1)),
      'top_right'   : (min(move[1],     9 - move[0]), ( 1, -1)),
      'bottom_left' : (min(9 - move[1], move[0]),     (-1,  1))
    }
    pos_flips = []
    for stop, (x, y) in directions.values():
      pos_flips.append(self._get_flip_positions(board, stop, move, x, y))
    return self._uniq(pos_flips)

  def _get_flip_positions(self, board, stop, move, x, y):
    pos_flips = []
    for i in range(1, stop):
      pos = (move[0] + x * i, move[1] + y * i)
      if board[pos] == EMPTY: break
      if board[pos] == PLAYER:
        for p in pos_flips:
          board[p] = PLAYER
        return pos_flips
      pos_flips.append(pos)
    return []

  def _uniq(self, arr):
    arr = reduce(lambda a, e: a if e in a else a + [e], arr, [])
    return [e for e in arr if e]

  def __str__(self):
    return str(self.board)

def won_game(record, display=False):
  othello = Othello()
  for move in record:
    if move == '0':
      break
    othello.make_move([int(move[0]), int(move[1])])
  PLAYER_count = 0
  for row in range(8):
    for col in range(8):
      if othello[row, col] == PLAYER:
        PLAYER_count += 1
  if PLAYER_count == 32:
    return None
  if display: print(othello)
  return PLAYER if PLAYER_count > 32 else OPPONENT

def make_othello_recoad():
  files = glob.glob('WTH/*')
  records = []
  for f in files:
    for l in open(f, 'rb'):
      l = l[16:]
      for e in [l[i: i + 68] for i in range(0, len(l), 68)]:
        records.append(e[8:])
  for record in records[0:10]:
    othello = Othello()
    for move in record:
      if move == 0: break
      othello.make_move([move // 10, move % 10])
    print(othello)

def main():
  unit = [100, 100, 64]
  model = Classifier(MLP(unit))
  chainer.cuda.get_device(0).use()
  model.to_gpu()
  optimizer = chainer.optimizers.Adam()
  optimizer.setup(model)
  chainer.serializers.load_npz('othello_model.npz', model)

  othello = Othello()
  for _ in range(60):
    print(othello)
    X1 = np.array([othello.board.cells], dtype=np.float32)
    X1 = cuda.to_gpu(X1, device=0)
    y = F.softmax(model.predictor(X1))
    y_ = int(y.data.argmax(1)[0])
    y = y_ // 8 * 10 + y_ % 8 + 11
    print("y = {}\n".format(y))
    move = int(input('move: '))
    othello.make_move((move // 10, move % 10))

if __name__ == '__main__':
  main()

File: othello.rb
require 'forwardable'

EMPTY = 0
PLAYER = 1
OPPONENT = 2
VALID = 3

DISK_MARK = {
  EMPTY => '-',
  PLAYER => 'o',
  OPPONENT => 'x',
  VALID => '.',
}

class Board
  extend Forwardable
  def_delegators :@cells, :each, :map, :map!
  attr_reader :cells

  def initialize(board=nil)
    if board.nil?
      @cells = Array.new(8) {[0] * 8 }
      @cells[3][4] = PLAYER
      @cells[4][3] = PLAYER
      @cells[3][3] = OPPONENT
      @cells[4][4] = OPPONENT
    else
      @cells = board.dup
    end
  end

  def []=(disc_row, disc_col, value)
    @cells[disc_col - 1][disc_row - 1] = value
  end

  def [](disc_row, disc_col)
    return EMPTY unless (1..8).to_a.include?(disc_row) &&
                        (1..8).to_a.include?(disc_col)
    @cells[disc_col - 1][disc_row - 1]
  end

  def dup
    @cells.map(&:dup).dup
  end

  def to_s
    '  ' + (1..8).map {|i| "#{i} " }.join + "\n" +
      @cells.each_with_index.reduce('') do |str, (row, i)|
        "#{str}#{i + 1} #{row.reduce('') {|a, e| a + "#{DISK_MARK[e]} " }}\n"
      end + "\n"
  end
end


class Othello
  attr_reader :board

  def initialize
    @board = Board.new
    @valid_moves = []
    update_valid_moves
  end

  def make_move(move)
    @valid_moves.each {|e| @board[*e] = EMPTY }
    flip_disks(@board, move)
    switch_turn
    update_valid_moves
    switch_turn if @valid_moves.empty?
  end

  def update_valid_moves
    @valid_moves.clear
    (1..8).each do |row|
      (1..8).each do |col|
        if @board[row, col] == EMPTY
          board = Board.new(@board)
          @valid_moves << [row, col] unless flip_disks(board, [row, col]).empty?
        end
      end
    end
    @valid_moves.uniq!
    @valid_moves.each {|e| @board[*e] = VALID }
  end

  def switch_turn
    @board.map! do |row|
      row.map do |cell|
        cell = cell == PLAYER ? OPPONENT : PLAYER if cell != EMPTY
        cell
      end
    end
  end

  def to_s
    @board.to_s
  end

  private
  def flip_disks(board, move)
    board[*move] = PLAYER
    right  = 1..(8 - move[0])
    bottom = 1..(8 - move[1])
    left   = 1..(move[0] - 1)
    top    = 1..(move[1] - 1)
    directions = {
      right:  [right,  [ 1,  0]],
      bottom: [bottom, [ 0,  1]],
      left:   [left,   [-1,  0]],
      top:    [top,    [ 0, -1]],
      top_left:     [[top, left].min_by {|r| r.size },     [-1, -1]],
      bottom_right: [[bottom, right].min_by {|r| r.size }, [ 1,  1]],
      top_right:    [[top, right].min_by {|r| r.size },    [ 1, -1]],
      bottom_left:  [[bottom, left].min_by {|r| r.size },  [-1,  1]]
    }
    directions.reduce([]) do |valid_moves, (dir, (range, pos))|
      valid_moves + get_flip_positions(board, range, move, *pos)
    end.reject(&:empty?).uniq
  end

  def get_flip_positions(board, range, move, x, y)
    range.reduce([]) do |pos_flip, i|
      pos = [move[0] + x * i, move[1] + y * i]
      break if board[*pos] == EMPTY
      if board[*pos] == PLAYER
        pos_flip.each {|p| board[*p] = PLAYER }
        return pos_flip
      end
      pos_flip << pos
    end
    []
  end
end


def serialize_board_str(board)
  board.cells.reduce('') do |str, row|
    str + row.reduce('') {|a, e| a + e.to_s }
  end
end

def main
  files = Dir.glob('WTH/*')
  records = []
  files.each do |f|
    data = File.open(f, 'rb').each_byte.map {|e| e }
    data[16..data.size - 1].each_slice(68).map {|e| e[8..67] }.each {|g| records << g }
  end
  encoding = lambda {|s| (s / 10 - 1) * 8 + s % 10 - 1 }
  output_data = []
  [records[0]].each_with_index do |record, i|
    othello = Othello.new
    record.each do |r|
      break if r == 0
      puts r
      output_data << serialize_board_str(othello.board) + " #{encoding.call(r)}"
      othello.make_move([r / 10, r % 10])
      puts othello
    end
    puts "#{i}: \n#{othello}" if i % 100 == 0
  end
  File.open('othello.data', 'w') {|f| output_data.each {|d| f.puts d } }
end

if __FILE__ == $0
  s = Time.now
  main
  puts "#{Time.now - s}s"
end
